From 26731a3059430c638a796e2d0904df26e7bc9553 Mon Sep 17 00:00:00 2001
From: Eran Cohen <ercohen@redhat.com>
Date: Tue, 12 Jan 2021 14:44:00 +0000
Subject: [PATCH] Add patch.service as a temporary hack

The patch.service is the temporary hack we use to complete the single
node installation. As such, it is not necessarily related to
bootstrap-in-place.

It will go away once cluster-etcd-operator and
cluster-authentication-operator support single node deployments.

Currently the service needs to run while cluster-bootstrap is running
(control plane is up).

We tried to move it to run after the reboot but noticed some race
https://bugzilla.redhat.com/show_bug.cgi?id=1905221#c21.

Since this is just for the POC and will be replaced by
cluster-etcd-operator and cluster-authentication-operator changes, we
have kept it in the bootstrap phase for now.
---
 .../bootstrap-in-place-post-reboot.sh                  | 10 ++++++++++
 pkg/asset/ignition/bootstrap/bootstrap.go              |  1 +
 2 files changed, 11 insertions(+)

diff --git a/data/data/bootstrap/bootstrap-in-place/opt/openshift/bootstrap-in-place/bootstrap-in-place-post-reboot.sh b/data/data/bootstrap/bootstrap-in-place/opt/openshift/bootstrap-in-place/bootstrap-in-place-post-reboot.sh
index 407164a47..a95285d40 100755
--- a/data/data/bootstrap/bootstrap-in-place/opt/openshift/bootstrap-in-place/bootstrap-in-place-post-reboot.sh
+++ b/data/data/bootstrap/bootstrap-in-place/opt/openshift/bootstrap-in-place/bootstrap-in-place-post-reboot.sh
@@ -46,6 +46,15 @@ function approve_csr {
   fi
 }
 
+function patch_ingress {
+  echo "patch ingress operator to run a single router pod"
+  while ! oc patch -n openshift-ingress-operator ingresscontroller/default --patch '{"spec":{"replicas": 1}}' --type=merge;
+  do
+    echo "Still waiting to patch ingress ..."
+    sleep 3
+  done
+}
+
 function wait_for_cvo {
   echo "Waiting for cvo"
   until [ "$(oc get clusterversion -o jsonpath='{.items[0].status.conditions[?(@.type=="Available")].status}')" == "True" ];
@@ -64,5 +73,6 @@ function clean {
 
 wait_for_api
 approve_csr
+patch_ingress
 wait_for_cvo
 clean
diff --git a/pkg/asset/ignition/bootstrap/bootstrap.go b/pkg/asset/ignition/bootstrap/bootstrap.go
index 4e1e88985..f9f87f46e 100644
--- a/pkg/asset/ignition/bootstrap/bootstrap.go
+++ b/pkg/asset/ignition/bootstrap/bootstrap.go
@@ -366,6 +366,7 @@ func (a *Bootstrap) addSystemdUnits(uri string, templateData *bootstrapTemplateD
 		"chown-gatewayd-key.service":      {},
 		"systemd-journal-gatewayd.socket": {},
 		"approve-csr.service":             {},
+		"patch.service":             {},
 		// baremetal & openstack platform services
 		"keepalived.service":        {},
 		"coredns.service":           {},
-- 
2.27.0

